### ISSUE TITLE
CPU stalls under qemu 7.1.0

### ISSUE THREAD

**Issue by anta5010** [OP-TEE/optee_os#5551]
Initially we noticed failures in meta-arm CI pipelines when qemu was updated from 7.0.0 to 7.1.0 in Yocto. We see issues when some types of machines fail to boot because optee linux kernel driver fails to initialise, other types boot ok, but CPU stalls detected when xtest is running.

  This is an example of NWd boot log when CPU stalls detected during optee linux kernel driver initialisation:
```
[    1.354550] optee: revision 3.18 (1ee64703)
[  337.412615] rcu: INFO: rcu_preempt detected stalls on CPUs/tasks:
[  337.413262] rcu:     2-...0: (1 GPs behind) idle=a27/1/0x4000000000000000 softirq=41/42 fqs=2127
[  337.413689]  (detected by 0, t=5252 jiffies, g=-1131, q=2 ncpus=4)
[  337.414074] Task dump for CPU 2:
[  337.414340] task:swapper/0       state:R  running task     stack:    0 pid:    1 ppid:     0 flags:0x0000000a
[  337.414907] Call trace:
[  337.415011]  __switch_to+0x120/0x1b0
[  337.415242]  0x0
```

  If required I can provide additional logs and other details from Yocto testings. But, an image booting issue can be reproduced without Yocto as well.
  
  I've:
- updated qemu ref to 7.1 in https://github.com/OP-TEE/manifest/blob/master/qemu_v8.xml#L26 and
- followed the build and run instructions from: https://optee.readthedocs.io/en/latest/building/devices/qemu.html#qemu-v8

  When I run `make run-only` the image boot hangs for ever.

1. NWd boot log ends with (full log is attached as `optee-nwd.log`):
```
EFI stub: Using DTB from configuration table
EFI stub: Loaded initrd from LINUX_EFI_INITRD_MEDIA_GUID device path
EFI stub: Exiting boot services...
```
2. SWd optee log is attached as `optee.log`

**Comment by jforissier** [comment]
Hi @anta5010,

I have bisected the issue down to commit https://github.com/qemu/qemu/commit/696ba3771894f7a0b233e634dc9d401330568e35 but I won't pretend I understand what this is supposed to do...

**Comment by anta5010** [comment]
Thank you @jforissier. Are you going to raise a qemu issue or contact the commit author?

**Comment by jforissier** [comment]
@anta5010 before doing that we should narrow the problem down I think. From the commit description I suppose TF-A or OP-TEE might be doing some invalid register access.

**Comment by anta5010** [comment]
@jforissier are you or someone else is planning to look at the issue deeper? I agree that it's not clear if it's qemu or OP-TEE or TF-A issue. I'm happy to help with tests in Yocto if required.

I'm sorry if I sound pushy, we're trying to understand what we need to plan for meta-arm Yocto layer. There will be a next Yocto release in a month time. Temporary for  qemuarm64 machines with OPTEE-os included I reverted the qemu7.1.0 update commit in meta-arm CI pipeline. But, it's not a solution which would be accepted into the release.

**Comment by jforissier** [comment]
There is a Data Abort with QEMU 7.1.0 that does not occur with 7.0.0. The CPU gets caught in `el1_sync_sp1`.
```
$ make -j10 CFG_CORE_ASLR=n
$ make run-only QEMU_EXTRA_ARGS="-d int" QEMU_SMP=1
[...]
(qemu) c
[...]
Taking exception 4 [Data Abort] on CPU 0
...from EL1 to EL1
...with ESR 0x25/0x96000004
...with FAR 0xffff80000800b990
...with ELR 0xe104c74
...to EL1 PC 0xe102200 PSTATE 0x3c5
```
https://esr.arm64.dev/#0x96000004 => read translation fault level 0

**Comment by jforissier** [comment]
Here is the outcome of my investigations so far. There are 3 separate issues (!):
1. The first one is what I described above: a data abort in the TEE core causes a endless loop in a TEE thread, which in turn causes the "RCU" messages from the kernel when several CPU cores are available (otherwise the boot just silently hangs). This bad behavior is introduced by QEMU commit [696ba37718](https://github.com/qemu/qemu/commit/696ba37718) ("target/arm: Handle cpreg registration for missing EL"), I have proposed a [patch](https://lists.gnu.org/archive/html/qemu-devel/2022-09/msg04515.html) upstream ("target/arm: mark SP_EL1 with ARM_CP_EL3_NO_EL2_KEEP").
2. The second issue is related to QEMU commit [39f29e599355](https://github.com/qemu/qemu/commit/39f29e599355) ("hw/intc/arm_gicv3: Use correct number of priority bits for the CPU"). It seems to be a kernel issue although I haven't identified the root cause yet. There is a workaround: use a compatibility mode, i.e., `-machine virt-7.0` instead of `-machine virt`.
3. The third issue is related to QEMU commit [78cb9776662a](https://github.com/qemu/qemu/commit/78cb9776662a) ("target/arm: Enable SME for -cpu max"). It also seems to be a compatibility issue with the kernel, and there is a similar workaround with `-cpu max,sme=off`.
```diff
diff --git a/qemu_v8.mk b/qemu_v8.mk
index c41219d..7e8ada8 100644
--- a/qemu_v8.mk
+++ b/qemu_v8.mk
@@ -406,7 +406,7 @@ QEMU_VIRT   = true
 QEMU_XEN       ?= -drive if=none,file=$(XEN_EXT4),format=raw,id=hd1 \
                   -device virtio-blk-device,drive=hd1
 else
-QEMU_CPU       ?= max,sve=off
+QEMU_CPU       ?= max,sve=off,sme=off
 QEMU_SMP       ?= 2
 QEMU_MEM       ?= 1057
 QEMU_VIRT      = false
@@ -430,7 +430,7 @@ run-only:
                -nographic \
                -serial tcp:localhost:54320 -serial tcp:localhost:54321 \
                -smp $(QEMU_SMP) \
-               -s -S -machine virt,secure=on,mte=$(QEMU_MTE),gic-version=$(QEMU_GIC_VERSION),virtualization=$(QEMU_VIRT) \
+               -s -S -machine virt-7.0,secure=on,mte=$(QEMU_MTE),gic-version=$(QEMU_GIC_VERSION),virtualization=$(QEMU_VIRT) \
                -cpu $(QEMU_CPU) \
                -d unimp -semihosting-config enable=on,target=native \
                -m $(QEMU_MEM) \
```

**Comment by anta5010** [comment]
@jforissier I can confirm that your proposed upstream patch fixes the CPU stalls issues we noticed in meta-arm Yocto CI pipelines.

**Comment by Andre-ARM** [comment]
@anta5010 I guess you are running with TF-A (and EDK-2) as firmware? I just had a similar issue, with both SVE and SME. By default the TF-A build disables those extensions for the QEMU platform, so skips the required EL3 initialisation. But the kernel still sees the feature bits, and per [booting.rst](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/arm64/booting.rst#n303) it assumes that the firmware did its EL3 duties.
This can be improved with setting `ENABLE_SVE_FOR_NS := 1` and `ENABLE_SME_FOR_NS := 1` in TF-A's [platform.mk](https://git.trustedfirmware.org/TF-A/trusted-firmware-a.git/tree/plat/qemu/qemu/platform.mk#n262), although the kernel still seems to hang later on (but only *after* some output). I am about to investigate that.

**Comment by jforissier** [comment]
>  This can be improved with setting `ENABLE_SVE_FOR_NS := 1` and `ENABLE_SME_FOR_NS := 1` in TF-A's [platform.mk](https://git.trustedfirmware.org/TF-A/trusted-firmware-a.git/tree/plat/qemu/qemu/platform.mk#n262), although the kernel still seems to hang later on (but only _after_ some output). I am about to investigate that.

Right, and setting these flags but enabling only SME not SVE on the QEMU command line (`-cpu max,sve=off,sme=on`) causes an assertion failure in QEMU:

```
qemu-system-aarch64: ../target/arm/helper.c:6121: sve_vqm1_for_el_sm: Assertion `sm' failed.
```

**Comment by jforissier** [comment]
Issue 3 [above](https://github.com/OP-TEE/optee_os/issues/5551#issuecomment-1259475043) is fixed by setting TF-A's `ENABLE_SME_FOR_NS=1` **and** `ENABLE_SME_FOR_SWD=1` **and** a QEMU [patch](https://patchwork.kernel.org/project/qemu-devel/patch/20221004072354.27037-1-jerome.forissier@linaro.org/). Then `-cpu max` can be used without `sve=off,sme=off`.

The fact that `ENABLE_SME_FOR_SWD` has to be enabled doesn't look right to me though, it should work even without it. More investigations needed...

**Comment by Andre-ARM** [comment]
@jforissier many thanks for looking into this!
So the QEMU patch seems to be needed, but just the type change (of `valid_mask`) and allowing `SCR_ENTP2`. We don't need to change all bits to ULL, the compiler will expand this appropriately if the target type is 64bits.
For TF-A: it (booting Linux through EDK-2) worked for me with just `ENABLE_SME_FOR_NS=1`, together with your QEMU patch. So can you double check? Or do you have some SME/TPIDR2 code in a secure payload?

**Comment by Andre-ARM** [comment]
@jforissier scratch that comment about 1U being sufficient, it indeed breaks when _clearing_ bits. It just happens to work without the change, because we don't clear bits in `value`, with the **current** code flow.

**Comment by jforissier** [comment]
> For TF-A: it (booting Linux through EDK-2) worked for me with just ENABLE_SME_FOR_NS=1, together with your QEMU patch. So can you double check? Or do you have some SME/TPIDR2 code in a secure payload?

You're right. OP-TEE is doing `mrs x0, fpcr` from SEL1 when `CFG_WITH_VFP=y`. For some reason I thought it was NS EL1 (Linux) that was causing the exception. One more point clarified, thanks for the hint.

**Comment by Andre-ARM** [comment]
For the records:
The TF-A QEMU platform build now defaults to including SME and SVE support, the respective [patch](https://review.trustedfirmware.org/c/TF-A/trusted-firmware-a/+/16994) has been merged.
The needed [QEMU patch](https://patchwork.kernel.org/project/qemu-devel/patch/20221004072354.27037-1-jerome.forissier@linaro.org/) has also been merged to target-arm.next.

**Comment by anta5010** [comment]
@jforissier FYI, in Yocto meta-arm CI pipelines we can successfully run op-tee with `machine virt` after your fix for the number 1 issue applied. But, we use `-cpu cortex-a57` not `-cpu max`.
